runtime/commonMain/src/kotlinx/serialization/SerialDescriptorBuilder.kt

```diff
@@ -0,0 +1,240 @@
+    return ArrayListClassDesc(typeDescriptor)
+}
+/**
+ * Creates descriptor for the type `List<T>`.
+ */
+@ImplicitReflectionSerializer
+public inline fun <reified T> SerialDescriptorBuilder.listDescriptor(): SerialDescriptor {
+    return ArrayListClassDesc(serializer<T>().descriptor)
+}
+
+/**
+ * Creates descriptor for the type `Map<K, V>` where `K` and `V` are types
+ * associated with [keyDescriptor] and [valueDescriptor] respectively.
+ */
+public fun SerialDescriptorBuilder.mapDescriptor(
+    keyDescriptor: SerialDescriptor,
+    valueDescriptor: SerialDescriptor
+): SerialDescriptor {
+    return HashMapClassDesc(keyDescriptor, valueDescriptor)
+}
+
+/**
+ * Creates descriptor for the type `Map<K, V>`.
+ */
+@ImplicitReflectionSerializer
+public inline fun <reified K, reified V> SerialDescriptorBuilder.mapDescriptor(): SerialDescriptor {
+    return HashMapClassDesc(serializer<K>().descriptor, serializer<V>().descriptor)
+}
+
+/**
+ * Creates descriptor for the type `Set<T>` where `T` is the type associated with [typeDescriptor].
+ */
+public fun SerialDescriptorBuilder.setDescriptor(typeDescriptor: SerialDescriptor): SerialDescriptor {
+    return HashSetClassDesc(typeDescriptor)
+}
+
+/**
+ * Creates descriptor for the type `Set<T>`.
+ */
+@ImplicitReflectionSerializer
+public inline fun <reified T> SerialDescriptorBuilder.setDescriptor(): SerialDescriptor {
+    return HashSetClassDesc(serializer<T>().descriptor)
+}
+
+private class SerialDescriptorImpl(
+    override val serialName: String,
+    override val kind: SerialKind,
+    override val elementsCount: Int,
+    builder: SerialDescriptorBuilder
+) : SerialDescriptor {
+
+    public override val isNullable: Boolean = builder.isNullable
+    public override val annotations: List<Annotation> = builder.annotations
+
+    private val elementNames: Array<String> = builder.elementNames.toTypedArray()
+    private val elementDescriptors: Array<SerialDescriptor> = builder.elementDescriptors.toTypedArray()
+    private val elementAnnotations: Array<List<Annotation>> = builder.elementAnnotations.toTypedArray()
+    private val elementOptionality: BooleanArray = builder.elementOptionality.toBooleanArray()
+    private val name2Index: Map<String, Int> = elementNames.withIndex().map { it.value to it.index }.toMap()
+
+    override fun getElementName(index: Int): String = elementNames.getChecked(index)
+    override fun getElementIndex(name: String): Int = name2Index[name] ?: CompositeDecoder.UNKNOWN_NAME
+    override fun getElementAnnotations(index: Int): List<Annotation> = elementAnnotations.getChecked(index)
+    override fun getElementDescriptor(index: Int): SerialDescriptor = elementDescriptors.getChecked(index)
+    override fun isElementOptional(index: Int): Boolean = elementOptionality.getChecked(index)
+
+    override fun equals(other: Any?): Boolean {
+        if (this === other) return true
+        if (other == null || this::class != other::class) return false
```

Comment:
I'm not sure we really have to prohibit comparison to other `SerialDescriptor` implementations